vortex-buffer/src/buffer_mut.rs

```diff
@@ -450,6 +450,56 @@ impl<T> FromIterator<T> for BufferMut<T> {
     }
 }
 
+/// Owned iterator over a [`BufferMut`].
+///
+/// Examples
+/// --------
+///
+/// ```
+/// use vortex_buffer::buffer_mut;
+///
+/// let mut a = buffer_mut![1u16, 2, 3];
+/// let b = buffer_mut![4u16, 5, 6];
+/// a.extend(b);
+/// assert_eq!(a.len(), 6);
+/// assert_eq!(a[3..6], [4, 5, 6]);
+/// ```
+pub struct BufferMutIterator<T> {
+    buffer: BufferMut<T>,
+    index: usize,
+}
+
+impl<T: Copy> Iterator for BufferMutIterator<T> {
+    type Item = T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index == self.buffer.len() {
+            None
+        } else {
+            let value = self.buffer[self.index];
+            self.index += 1;
+            Some(value)
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let remaining = self.buffer.len() - self.index;
+        (remaining, Some(remaining))
+    }
+}
+
+impl<T: Copy> IntoIterator for BufferMut<T> {
```

Comment:
I would expect the into_iterator to actually consume the buffer instead of copying values

Reply:
otherwise you can make &T as a return value and the caller can copy

Reply:
I don't understand this, doesn't deref and derefmut give you &T and &mut T iterators already?

Reply:
you're correct that it would give you that but I didn't realize deref impls exist before making this comment

Reply:
I want this to compile:
```
    let buf: BufferMut<T> = values.into_buffer_mut::<T>();
    let mut new_buf: BufferMut<T> = BufferMut::<T>::with_capacity(buf.len() + 1);
    new_buf.push(T::zero());
    new_buf.extend(buf);
```
But it doesn't because `buf` does not implement `IntoIterator<Item = T>`. I somehow want to convince Rust to just memcpy the old buffer into the new buffer.

Reply:
You should check out how this is handled for Vec - spoiler alert, you want to be able to introspect passed iterator and extract the buffer back from it.

Reply:
I can punt on all of this and use a for loop because I need to increment each code by one anyway. I'll check out the Vec implementation though because I'm definitely missing something about how to use Rust.